### ISSUE TITLE
Maturing out the package (Path to DataStructures.jl v1.0)

### ISSUE THREAD

**Issue by oxinabox** [JuliaCollections/DataStructures.jl#479]
What do we need to do before we would be happy to tag DataStructures.jl v1.0?
I mean, we are one of the most common dependencies in the package ecosystem,
and julia 1.0 is out.

My thoughts;
 
 - [ ] Search the code base for anything written in outmoded styles (e.g. #472)
 - [ ] Get rid of anything deprecated #477 (but that will need to be done again after)
 - [ ] Split up the package #310, but noting that all the packages sliced out that way will want to hit 1.0 before (or at the same time as) DataStructures.jl does.

It hurts people when we break things, and while we are good at deprecating,
getting on the the "strict" side of semver is something we need to do for the sake of the ecossytem

**Comment by oxinabox** [comment]
People keep asking when will DataStructures 1.0 come out.
The answer is as soon as we completely audit the package to make sure there is nothing that needs to be removed/corrected to follow modern julia conventions.
And have actioned those corrections.
A lot of the code in the package is pretty old, and follows e.g. conventions from julia 0.3.
We need to identify and correct these.

I will post a list of files at the bottom of this post.
Then people can go through this and use the checklist to make sure everything is good, one file at a time.




How to audit a file to check if 1.0 ready
---

 - Answer the question of **should we even keep this in the package**. E.g. no data structure that somone just kind of made up (like the now removed `Classifier Collection)
 - **Type Names Correct/Sensible:** for example make sure we don't have a type called `MagicVectors` that is a subtype of vector. That should be called `MagicVector` (singular).
 - **No constructors taking types as positional arguments.** For example there should not be `MagicVector(::Type{T}) where T`, instead there should just be `MagicVector{T}()`.
 - **No functions that duplicate constructors**. For example there should not be a function `magic_vector(xs)`, only `MagicVector(xs)`.
 - **No functions that are improperly cased and pretend to be constructors**. 
    - E.g. no `MagicVector(xs) = Vector(xs) .* "Magic"`, that should just be a normal function `magic_vector(xs) =  Vector(xs) .* "Magic"`
    - E.g. no `MagicVector(xs)=Vector{Magic}(xs)`, that should be a `const` type alias `const MagicVector=Vector{Magic}`
 - **No exported functions that could be just done with overloads of `Base`**. E.g. don't need `top`, `peak`, `head` all of which act the same as `first` but on a particular data structure.
 - **No internal helpers exported**. If something isn't useful to the end user it should not be exported. In general The vast majority of the public API should be constructor or overloads of `Base` functions. Thus any export should be looked at closely.
 - `push!` and `empty!` and `delete!` should return the collection. (this is relatively minor as it would be a bug if they didn't so it isn't breaking for use to fix it later.)


List of files that need to be checked
--------------------------------------
(Checking each will also require checking the top-level file for relevant exports)


- [ ] accumulator.jl
- [ ] default_dict.jl
- [ ] robin_dict.jl
- [ ] dict_support.jl

- [ ] trie.jl
- [ ] balanced_tree.jl
- [ ] circ_deque.jl
- [ ] circular_buffer.jl
- [ ] dibit_vector.jl

- [ ] disjoint_set.jl
- [ ] fenwick.jl
- [ ] int_set.jl
- [ ] sparse_int_set.jl

- [ ] list.jl
- [ ] multi_dict.jl
- [ ] mutable_list.jl

- [ ] deque.jl
- [ ] queue.jl
- [ ] stack.jl

- [ ] sorted_dict.jl
- [ ] sorted_multi_dict.jl
- [ ] sorted_set.jl
- [ ] tokens.jl
- [ ] tokens2.jl


- [ ] priorityqueue.jl
- [ ] heaps.jl
- [ ] heaps/arrays_as_heaps.jl
- [ ] heaps/binary_heap.jl
- [ ] heaps/minmax_heap.jl
- [ ] heaps/mutable_binary_heap.jl

- [ ] container_loops.jl
- [ ] delegate.jl


Higher level process:
---------------------

0.19 will be the last release before 1.0.
0.19 will be the same as 1.0 but with deprecations.
1.0  will come out the same day as 0.19

We will stick to making patch releases on 0.18 with deprecations while we can.
But once we can't we will swap master to 0.19-DEV, and say on that until we can release it and 1.0.
That might take a 6 or more months; depends how many (if any) people are about to spend time helping with the audit.
If necessary we will have a backport branch for 0.18 bug-fixes

Once @eulerkochy's GSOC work is complete we will not accept any new data structures until after 1.0 released.

**Comment by StephenVavasis** [comment]
I would like to make some small revisions to the documentation for sorted containers (see e.g. https://github.com/JuliaCollections/DataStructures.jl/issues/528).  I don't know how to do this, i.e., I don't know which file I should edit because I don't understand how the docs are organized and populated.  Any pointers would be appreciated.

**Comment by nickrobinson251** [comment]
> No exported functions that could be just done with overloads of Base

How about [`reverse_iter`](https://juliacollections.github.io/DataStructures.jl/latest/stack_and_queue/)? Can this just be `Iterators.reverse`? The only issue i see is that `reverse_iter(s::Stack)` returns a `DequeIterator`, not a `Iterators.Reverse`, but idk if Iterator's API is really supposed to guarantee `Iterators.reverse(::T) -> Iterators.Reverse{T}` or not.

**Comment by nickrobinson251** [comment]
> * No exported functions that could be just done with overloads of Base.
>
> * No defining methods in ways that disagree with Base

I think this suggest we should make a call on which side this falls: https://github.com/JuliaCollections/DataStructures.jl/issues/296

**Comment by StephenVavasis** [comment]
Could anyone explain in more detail the above-mentioned recommendation regarding the `insert!` function?  In particular, does the recommendation mean that `insert!(ss, k)`, where ss is a SortedSet and k is a new key, should be deprecated?  Note that `push!` does not replicate this functionality because `insert!(ss,k)` returns a bool to indicate whether the item was already present and a semitoken indexing the newly inserted item, whereas `push!(ss,k)` returns the container.

**Comment by oxinabox** [comment]
> Could anyone explain in more detail the above-mentioned recommendation regarding the insert! function?

Base defines `insert!` as:


>   `insert!(a::Vector, index::Integer, item)`
> 
>   Insert an item into a at the given index. index is the index of item in the resulting a.
> 

and it returns the collection.

> In particular, does the recommendation mean that insert!(ss, k), where ss is a SortedSet and k is a new key, should be deprecated?

yes, it doesn't match the description of the Base function.

> Note that push! does not replicate this functionality because insert!(ss,k) returns a bool to indicate whether the item was already present and a semitoken indexing the newly inserted item, whereas push!(ss,k) returns the container.

Yeah, so in that case we can't deprecate it for `push!`.
We would need to make some other plan.
Like `st_push!` or something? or a submodule that defines a bunch of function names for things that return subtokens?
Maybe a case could be made for `push!(::Type(SemiToken),::SortedSet, x)`  or something?

**Comment by StephenVavasis** [comment]
Here is a more serious conflict with `Base`: according to the Base docs, `pop!(s)` is supposed to return the last item in `s` if `s` is an ordered collection. However, `pop!` on the sorted containers currently returns the first item.  Which is the lesser of two evils, a breaking change for the package or doing the opposite of the `Base` documentation?

**Comment by StephenVavasis** [comment]
Maybe deprecate `pop!(s)` for sorted containers to `popfirst!(s)`, and introduce new functionality `poplast!(s)` (the latter not present in Base)?  Then, several years in the future, deprecate `poplast!(s)` to just `pop!(s)`?

**Comment by oxinabox** [comment]
> Maybe deprecate pop!(s) for sorted containers to popfirst!(s), and introduce new functionality poplast!(s) (the latter not present in Base)? Then, several years in the future, deprecate poplast!(s) to just pop!(s)?

yeah this, sounds right.
but I would argue for putting a deprecation warning in `poplast!` in 0.19 as it is defined saying this will just be `pop!` in 1.0, which will be released not years in the future but rather days.

**Comment by StephenVavasis** [comment]
Doesn't this approach for deprecating `pop!(s)` mean that everyone who upgrades from 0.18 to 1.0, skipping 0.19, will encounter a breaking change that is also silent?

With regard to renaming `insert!`: since `insert!` currently has a different API for each of the three containers, how about if we choose the three names `ss_push!`, `sd_push!` and `smd_push!` ?

**Comment by oxinabox** [comment]
> Doesn't this approach for deprecating pop!(s) mean that everyone who upgrades from 0.18 to 1.0, skipping 0.19, will encounter a breaking change that is also silent?

yes, that is what they get for skipping versions. But we will write release notes etc.

> With regard to renaming insert!: since insert! currently has a different API for each of the three containers, how about if we choose the three names ss_push!, sd_push! and smd_push! ?

Lets split this off to another issue or a PR and we can think about our options.

**Comment by StephenVavasis** [comment]
OK.  I am currently working on a major overhaul of the sorted containers to address a number of open issues at once. I will include these new names in my forthcoming PR; and then you or another reviewer can comment on them.

**Comment by StephenVavasis** [comment]
With regard to attaching proper meanings to `push!`, `pop!`, and `insert!`: Kevin Squire @kmsquire wrote a long message on discourse about how these functions are somewhat muddled in `Base`:

https://discourse.julialang.org/t/taking-push-and-pop-seriously/34326

He proposed a more principled way to implement them in Base, but, as far as I can tell, did not get any traction for this proposal.  Presumably this is because it is now too late to make changes to Base.  If we try to follow Base conventions in the DataStructures package, then we will end up with the same muddled semantics.

I'd also like to point out open issue https://github.com/JuliaCollections/DataStructures.jl/issues/403 which is basically a request for clearer semantics for these functions.

**Comment by Suavesito-Olimpiada** [comment]
One thing I've found is that sometimes I would like to use special ordering that _I do not want to define with `isless`_, maybe because it would be type piracy or simply does not make sense outside the context of the algorithm implemented. Would it be possible for the data structures that need an ordering function to always accept something like `lt` for `sort`?

**Comment by StephenVavasis** [comment]
This is possible-- the sorted containers are parameterized by an order function that you can customize.  Please see the documentation for an example in which strings are ordered by case-insensitive comparison.

**Comment by stevengj** [comment]
I don't understand why you're setting such a high bar on 1.0.

Lots of packages depend on DataStructures.jl *right now*, and by using pre-1.0 version numbers means every non-bugfix released is treated as "breaking" by the package system whether or not it is backwards compatible.

Version numbers are cheap.   Can you just make the next minor release 1.0?  If you ever want a major breaking change you can just mark the package as 2.0 in the future.

**Comment by oxinabox** [comment]
> I don't understand why you're setting such a high bar on 1.0, which has caused this issue to stay open for five years.

When this was open @ChrisRackauckas had gotten made at us for releasing 0.18 of DataStructures.
So we said no more breaking releases til we can get to 1.0.
But yes, when we said that *I* had more time, and we had more people actively working on the package.
So was hoping to do it in much less time.

The current `#master` branch is breaking and does have a few good improvements.
I would not be opposed to cutting a 1.0 branch from that.
It would decrease my work, because wouldn't have to keep backporting bug fixes.

But it would mean a lot of packages would need to go and update.
On the other hand, we do have better tooling like CompatHelper now to ease that pain.

**Comment by ChrisRackauckas** [comment]
You went from v0.12 to v0.18 in the span of 2 years (some of those shouldn't've been marked as breaking probably?), and now this one is 5 years in the making. There's a middle ground.

**Comment by fingolfin** [comment]
I would generally support taking the plunge and getting master release as 1.0...

Except that of course due to `semver`'s limitation it will probably be a long and arduous treck to get everyone switched to 1.0: Essentially for package like this with soooo many other package depending on it, a breaking release risks re-creating the python 2->3 problem each time.

To avoid this, I think the most critical bit are thorough release notes that describe what exactly are the breaking changes, ideally including recipes for how to deal with them.

And possibly encouraging downstream packages (i.e. packages depending on DataStructures.jl) to try and support both 0.18.x and 1.0.x in parallel, at least for a while, to minimize suffering ("my dependency A still requires DS 0.18, but dependency B only support DS 1.0, so I need to keep B at an older version until they change that").

**Comment by AzamatB** [comment]
Can we tag a new release of DataStructures.jl? I need to use a non-string version of the Tries, which has been [added](https://github.com/JuliaCollections/DataStructures.jl/pull/759) ~4 years ago, but it is still not in the latest version for some reason. It does not need to be 1.0. It can be just 0.19 or something.

**Comment by oxinabox** [comment]
We probably should. It would have been nice to hold off to proper 1.0 was ready but it isn't going to happen anytime soon.

Maybe just 0.19

**Comment by jonas-schulze** [comment]
See #769

**Comment by stevengj** [comment]
> Maybe just 0.19

semver will also treat this as breaking, so I don't see any advantage in marking a new release as 0.19 vs. 1.0.  Whereas 1.0 has the advantage of providing strictly more information to downstream users for subsequent 1.x.y releases.

**Comment by StephenVavasis** [comment]
There was a plan with @oxinabox regarding the corner of this package that I maintain (the sorted containers) to, in one release, deprecate some of the functionality that corresponds to Base functionality but with different names or API details, and then, in a subsequent release, drop those inconsistent names or APIs entirely.  See: https://github.com/JuliaCollections/DataStructures.jl/pull/787  This conversation took place three years ago, so my recollection is a bit fuzzy.

**Comment by oxinabox** [comment]
Honestly i am infavor of option 3.
But yes it is just lacking someone to do that work.
E.g. even for 3 we would want to prep some release notes,
and at this point it would require detective work.

Re: point 1.
Most of the non-breaking stuff has been backported to the 0.18.
I have been fairly diligent about that.

**Comment by stevengj** [comment]
Yes, please just make a release from the current repo state, and call it 1.0 for the reasons noted above.

**Comment by PallHaraldsson** [comment]
I see lots of support for my (mostly redundant with later discussion here like):
https://github.com/JuliaCollections/DataStructures.jl/issues/479#issuecomment-2322604225

If people are hesitant of 1.0 right now, pushing it off for more years, potentially, because something is less perfect, can those datastructures be moved to UnstableDataStructures.jl with pre-1.0 version number that depends on this new 1.0 (since a breaking version number, dropping those lesser used/developed ones)?

**Comment by oxinabox** [comment]
So the reason this hasn't made much progress in last 7 years , despite the fact that is the plan is fairly straight forward.
But actually doing it requires 2 or 3 people to sit down who really care to do it, and the review each other's code.

I made a start in https://github.com/JuliaCollections/DataStructures.jl/pull/837
but felt like i didn't get enough feedback to really come to conclusion on what is best.

Finding 2-3 people who really care and getting them to dedicate a week to this is hard.
_I_ certainly have better things to do with my time.
